Short passwords with PBKDF2 mode working

mposolda opened the pull request:
Closes #14314

@stianst @pedroigor Can I please ask anyone from you to review this?

mposolda committed:
Short passwords with PBKDF2 mode working
Closes #14314

pedroigor reviewed (commented):
In addition to password length do we need to consider the salt length too? See https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3152.pdf.

I'm also wondering if instead of padding we should just enforce the constraint around the password length? Isn't that the purpose of using BCFIPS, to be more constrained?

stianst reviewed (commented):
I'm not really liking this to be honest.

Padding with `0` is not great from an entropy perspective, another issue is that we are now generating hashed passwords that are incompatible between FIPS and non-FIPS mode. For example if FIPS is first enabled, then later disabled for some reason. Wouldn't this also mean if FIPS is enabled all current passwords would be invalid? 

Is this an issue in BC? Did you try to use PKBDF2 from other security providers?

I find the statement `In approved mode the module enforces that any password used must encode to at least 14 bytes (112 bits) and
that the salt is at least 16 bytes (128 bits) long. The iteration count associated with the PBKDF should be as
large as practical.` from BC a bit strange. What do they mean by it really? Depending on how you `encode` a password the number of bites will be different. For example UTF-8, UTF-16, Base64, there's plenty of ways a password can be encoded with more bytes, without adding any additional entropy.

In addition to the above I also find it slightly weird that we have a "CryptoProvider" that pads passwords in general ;)

I think this one needs a bit more digging, which is a PITA as wasting time on this just feels like exactly that waste.

mposolda commented:
@pedroigor @stianst Thanks for the pointers! My current thinking was to use this as a step 1 and then as step 2 figure what are the requirements for the passwords in BCFIPS (I've tried to dig a bit in last days, but didn't yet found anything concrete).

Sent the email to keycloak-fips ML about password requirements. Also yesterday asked the community member, who uses FIPS, about his deployment on the GH issue linked to this PR. Maybe worth to start discussion in community for bigger feedback though...?

Answers to your questions:
- Salt : Our Pbkdf2PasswordHashProvider uses 16 bytes salt and it always used salt of that size in previous KEycloak versions. BCFIPS prescribes 16 bits, so salt length shouldn't be an issue in BCFIPS approved mode. We can consider increasing default size of the salt though for the security purpose (option of Pbkdf2 java provider maybe?), but maybe also as a follow-up?

> For example if FIPS is first enabled, then later disabled for some reason. Wouldn't this also mean if FIPS is enabled all current 
> passwords would be invalid?

Passwords created without padding works when you use padding with "\0". In other words, passwords created in previous Keycloak versions should still work. I even have a automated test for that in this PR.

> Is this an issue in BC? Did you try to use PKBDF2 from other security providers?

On FIPS OpenJDK, there are not any providers for PBKDF2 besides BCFIPS. However, even if there is some PBKDF2 provider without this restriction, which allows short password, then using this provider with short passwords is not any safer than using BCFIPS with padding, right?

> For example UTF-8, UTF-16, Base64, there's plenty of ways a password can be encoded with more bytes, without adding any 
> additional entropy.

Yes, we can use for example UTF-8 or something for more entropy. However this exactly won't work for backwards compatibility for passwords created in previous versions, when padding works.

One possible way will be to use UTF-8 together with Base64 by default (It needs to be by default even for non-fips due the migration of KEycloak deployment between FIPS and non-FIPS environment). However at least for the verification of the old short password, we still need to use padding. After verification, the password can be updated to use UTF-8 + Base64 (similar like we are doing for hashing iterations). For this, we can add new property to PasswordCredentialData to specify if it was created before with UTF-8 and Base64. WDYT?

My current thinking is to:
1) Use this PR as is and then figure password policies as a follow-up to make sure that passwords are long enough

2) Update this PR to use bigger entropy for short passwords by default. In other words, when short passwords arrives, it will need to be encoded to UTF-8 / Base64 - possibly multiple times until the length is at least 14 characters. Add the flag to PasswordCredential for this, so Keycloak knows when it should use this entropy or not during password verification. The padding will still need to be used for password verification of old passwords with BCFIPS approved-mode (I am not really see how to avoid padding if we want to be backwards compatible).

In both cases, figure best practices for FIPS (password policies etc) as a follow-up.

stianst commented:
Thanks,

As long as we make sure we have a follow-up on this one I'm okay.

One suggestion though with regard to the implementation. I find it's a bit odd that the CryptoProvider pads the password, while in fact the issue really lies in the PBKDF2 hashing. So was wondering about if a slightly different approach would make sense.

What about instead we have a config option on the PBKDF2 hashing provider that let's you enable/disable padding for short passwords. Could just make it a server config since we don't really have the nicest way to configure hashing providers and it really should be a server-wide thing anyways. We could perhaps have this option enabled by default in BC-FIPS mode.

It feels more natural to me to solve this problem within the hashing provider, and is also then something a bit more visible that can be enabled/disabled. Would also be a bit more natural if additional hashing providers are made available that don't have this problem.

mposolda commented:
@stianst Yes, so there is follow-up task for password policies setup in FIPS: https://github.com/keycloak/keycloak/issues/14457

For this one, I've removed the changes from `CryptoProvider` and directly updated Pbkdf2PAsswordHashProvider to always use padding for shorter passwords. As I've mentioned, there is not any side-effect and passwords created before without padding works fine, so backwards compatibility is also ok. So we can avoid another config option and just always do this padding by default? It looks that Pbkdf2 impl is doing something like this under the covers, so this is just to move it to Keycloak to overcome the BCFIPS restrictions.

I've used separate commit, so you can see the changes since last update (This can be squashed during merged)

pedroigor commented:
I'm not sure if padding by default is the best way to go. If I got the problem correctly, you are trying to make it easier to transition from non-FIPS to FIPS-approved mode.

I think the decision to do this form of padding should not come from us as it depends on your security requirements. For instance, does it make sense to suddenly switch to FIPS or deployments should be using FIPS from day 1?

If I want to avoid breaking users using a short password, should I explicitly choose to pad passwords and configure the provider to use padding?

trixpan commented:
@mposolda sorry I could not reply before. I will reach out offline with some extra details but overall we don't use keycloak for storing user passwords.

@pedroigor has a valid point. I don't think we should worry about supporting migration to and from FIPS without full redeployment. It is not just the passwords but the keys as well. The idea of using FIPS is that by using a certified crypto provider the keys are built to the required standard. How are you going to flag a key as built without FIPS from those with FIPS?

I have the hard view that a system either is FIPS from day zero or it is not FIPS compliant. I don't believe FIPS itself has requirements around that but it is just basic hygiene really. A system cannot be deemed trustworthy if the controls are turned on and off at the operator's whims (and reason why higher level FIPS compliance requires temper resistance of devices).

I do have some points to make:

Most places required to use FIPS will also be required to have policies enforcing good passwords as part of the other controls associated with that environment (e.g. FedRAMP High requires [at least](https://help.ivanti.com/ht/help/en_US/ISM/2018.3/user/Content/Configure/FedRamp/FedRAMP_Configurations.htm) [15 characters](https://help.salesforce.com/s/articleView?id=000339120&type=1)).

So if you accept my suggestion I would simply say, instead of working around by padding we should simply detect FIPS usage and force password policies to enforce 15 min characters (and avoid using any sort of padding or magic around creds).

Maybe @JPercivall may also share some of his views.

mposolda commented:
@trixpan @pedroigor This is good point that most places required to use FIPS will also be required to have policies enforcing good passwords. On the other hand, the FedRamp recommendation of 15 characters is not officially enforced by FIPS itself, right? Do you know what are official FIPS 140-2 requirement around passwords?

Another question is, if we can really support FIPS just for those, who use it from day-1 or if we need to support also those, who migrate from non-FIPS environment (where the users might already have passwords shorter than 14 characters). I don't have an answer to this, but I suppose that also those, who are migrating. On the other hand, they already might have some option (EG. in case that realm has "Forget password", users can be enforced to change their password to match password policies). So agreee that in that case, the padding may not be needed.

Another thing is how to deal with enforcing password requirements. Keycloak has mechanism of password policies. So to enforce FIPS requirement, password policies can be:
- predefined: When new realm is created, the password policies are automatically created on the realm. This gives administrator possibility to remove them. If we use BCFIPS without padding, this option won't easily work as it is just not allowed to have shorter passwords than 14 chars.
- enforced: Password policies are enforced at runtime. So for example administrator won't have possibility to use password policy to allow shorter password than 14 characters (but he should still have possibility to use stronger password).

The concept of "enforced" password policies is not yet available in Keycloak, but can be possibly added as part of the FIPS work. The "enforced" policies will require some support in the admin console and hence it is a bit of the additional effort though.

stianst commented:
I'm curious what the real requirements are around passwords in FIPS. I haven't been able to find anything concrete here.

With regard to password policies I would probably just leave this as a documentation issue. We can mention what the minimum password policies are and just require folks to configure it accordingly.

stianst commented:
I think it's a valid use-case to be able to convert a non-FIPS installation to a FIPS installation, which probably isn't going to be all that trivial perhaps. Not sure how best we'd achieve that. I could imagine two approaches here: 

1. Document what is required so folks can update keys/password policies etc. prior to enabling FIPS
2. A reporting only mode for FIPS that still allows shorter keys/passwords/insecure algorithsm, but allows reporting on it, so non-compliant stuff can be changed prior to enabling enforcing mode

mposolda commented:
@stianst I am also not able to find anything concrete around password policies in FIPS 140-2. Asked on various places (keycloak-fips ML etc), but not have any clear answer. Fact is, that most deployments use some kind of password policies.

So how about this in this PR: We can add option on Pbkdf2PasswordHashProvider for enable padding. By default, it won't be enabled. The configurable option will be int and it would be count of padding characters. So guys, who want to use BCFIPS with approved mode will have options like:
- Enable padding to 14 characters
- Not enable padding and instead make sure that realm password policy is configured to 14 characters

There can be combination of both options for those, who want to migrate from non-fips environment (as with password policy alone, the guys with existing shorter passwords migrated from previous version won't be able to authenticate with those shorter password when BCFIPS approved mode is enabled).

These options will be documented. We will also need to continue digging for figure the requirements around passwords in FIPS (if any requirements exists at all) and document those as a follow-up. WDYT?

mposolda committed:
Add config option to Pbkdf2 provider to control max padding

mposolda commented:
@stianst I've updated PR as we discussed and as specified in the last comment. Can you please re-review? I've used separate commit for the updates, so it is easily possible to check what was updated. I am planning to squash during merge.

stianst commented on server-spi-private/src/main/java/org/keycloak/credential/hash/Pbkdf2PasswordHashProvider.java:
```diff
@@ -132,4 +136,14 @@ private SecretKeyFactory getSecretKeyFactory() {
         }
     }
+
+    private String pbkdf2PasswordPadding(String rawPassword) {
+        if (rawPassword.length() < maxPaddingLength) {
+            int nPad = maxPaddingLength - rawPassword.length();
+            String result = rawPassword;
+            for (int i = 0 ; i < nPad; i++) result += "\0";
+            return result;
+        } else
+            return rawPassword;
+    }
```
Would have liked to see something testing this padding stuff. What about making it a static method, extracting into a separate class, and adding a unit test?

mposolda replied:
There is unit test `FIPS1402Pbkdf2PasswordPaddingTest` in this PR. It tests various scenarios like backwards compatibility (password without padding is possible to verify with padding) and also check that it works to switch between non-approved and approved mode for password verifications etc.

It is in the `crypto/fips1402` module, and hence only in the FIPS specific module, however I consider this OK due the fact:
- Padding will be most likely used especially by guys in the FIPS mode, if they cannot rely just on password policies (for example due the backwards compatibility and requirement of verification of old passwords)
- These tests fare executed during our regular GH actions build, which runs the unit tests on the `crypto/fips1402` module.

Isn't it sufficient?

stianst replied:
The tests now as I understand it checks that the password is compatible with padding on/off, and that it passes for the BC provider.

However, it would be nice to see a simple unit test (or something) that just verifies the padding of the string works as expected before the hashing is performed. Could for instance be that it actually pads it to a few characters longer than it should, etc.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/PasswordHashingTest.java:
```diff
@@ -224,6 +225,18 @@ public void testPbkdf2Sha512() throws Exception {
         assertEncoded(credential, "password", credential.getPasswordSecretData().getSalt(), "PBKDF2WithHmacSHA512", 30000);
     }
 
+    @Test
+    public void testPbkdf2Sha256BackwardsCompatibility() throws Exception {
+        setPasswordPolicy("hashAlgorithm(" + Pbkdf2Sha256PasswordHashProviderFactory.ID + ")");
+        String username = "testPbkdf2Sha2562";
+        createUser(username);
+
+        PasswordCredentialModel credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
+
+        // Encoded without padding still works and should give same results as originally padded password
+        assertEncoded(credential, "password", credential.getPasswordSecretData().getSalt(), "PBKDF2WithHmacSHA256", 27500);
+    }
+
```
Yes, this is good catch :-)

I think this is just thombstone during development of various versions of this PR (in some previous version, the padding was enabled by default when I wanted to test backwards compatibility, but now it is not enabled and this method is completely same as the one you pointed).

I will remove this method from this PR.

stianst replied:
Wonder if it would be nice to have a non-FIPS test that checks with padding enabled?

mposolda commented on server-spi-private/src/main/java/org/keycloak/credential/hash/PasswordHashSpi.java:
```diff
@@ -26,14 +26,15 @@
  */
 public class PasswordHashSpi implements Spi {
 
+    public static final String SPI_NAME = "password-hashing";
```
+1, updating as you suggested

mposolda committed:
Update according to PR review - more testing for padding and for non-fips mode

stianst commented on server-spi-private/src/main/java/org/keycloak/credential/hash/PasswordHashSpi.java:
```diff
@@ -26,14 +26,15 @@
     @Override
     public String getName() {
-        return "password-hashing";
+        return SPI_NAME;
```
```suggestion
        return NAME;
```

mposolda commented:
@stianst Thanks for the review. I've updated PR to:
- Adding separate static class with util method for padding. This is consumed by Pbkdf2PasswordHashProvider. Also added unit test for that
- Changed the test in the integration testsuite to test with enabled padding for Keycloak server. This should ensure that we test with both FIPS and non-FIPS mode.
- Renamed constant on all places (NAME vs previous SPI_NAME)

Again, used separate commit for the latest changes. Can you please re-review?
